direktiv_api: workflow/v1

description: Run a Docker Hub Python container without any changes

functions:
  # This is a simple Python container from hub.docker.com: docker pull python:3.9.18-slim-bookworm
  - id: python
    image: python:3.9.18-slim-bookworm
    size: medium
    type: knative-workflow
    cmd: /usr/share/direktiv/direktiv-cmd
 
states:
  # 
  # Extract the files from the multipart document sent from the request
  #
  - id: extract-multipart
    type: action
    action:
      function: python
      files: 
      - key: /run-docker-container/code/extractFiles.py
        scope: file
        as: extractFiles.py
      input: 
        data:
          commands:
          - command: pip install requests_toolbelt
          - command: python3 extractFiles.py -b jq(.body.data) -c 'jq(.headers."Content-Type"[0])'
    transform: 'jq( { output: ( .return[1].Output | fromjson) })'
    transition: store-code
    catch:
    - error: "*"
      transition: exception-catch-handle

  - id: store-code
    type: setter
    variables:
    - key: jq(.output[0].fname)
      scope: workflow
      mimeType: text/plain
      value: 'jq( .output[0].content )'
    - key: jq(.output[1].fname)
      scope: workflow
      mimeType: text/plain
      value: 'jq( .output[1].content )'
    transition: run-code

  - id: run-code
    type: action
    action:
      function: python
      files: 
      - key: jq(.output[0].fname)
        scope: workflow
        as: jq(.output[0].fname)
      - key: jq(.output[1].fname)
        scope: workflow
        as: jq(.output[1].fname)
      input: 
        data:
          commands:
          - command: pip install -r jq(.output[0].fname)
          - command: python3 jq(.output[1].fname)
    transform: 'jq( { output: ( .return[1].Output | fromjson) })'
    catch:
    - error: "*"
      transition: exception-catch-handle

  #
  # Handle the errors returned
  #
  - id: exception-catch-handle
    type: noop
    log: jq(.)